Deprecate building saveable records from read-only relations

eugeneius opened the pull request:
Currently records loaded from read-only relations are read-only, but records built from them are not. It shouldn't be possible to modify the data underlying a read-only relation, including adding new records.

kamipo commented:
Do not need a deprecation cycle?
This apparently changes the published behavior (readonly is to be applied to loaded records from database, new record allows save) that existing apps may rely on for a long time.

eugeneius commented:
Yes, you're right! I updated the patch to add a deprecation instead. I also reverted the strict loading part, since the two changes are no longer similar.

eugeneius commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -852,7 +852,7 @@ def none! # :nodoc:
     end
 
-    # Sets readonly attributes for the returned relation. If value is
+    # Sets the returned relation to readonly mode. If value is
```
I kept this docs change because the current version is misleading; "readonly attributes" are a different feature:

https://api.rubyonrails.org/classes/ActiveRecord/ReadonlyAttributes/ClassMethods.html

kamipo commented on activerecord/test/cases/readonly_test.rb:
```diff
@@ -44,6 +44,30 @@ def test_find_with_readonly_option
+  def test_build_with_readonly_option
+    assert_not_predicate Developer.all.new, :readonly?
+    assert_deprecated do
+      assert_not_predicate Developer.readonly.new, :readonly?
```
What should users do to avoid this warning?

eugeneius replied:
I considered suggesting `readonly(false)` in the deprecation message, but that won't necessarily work for read-only associations since the new record then wouldn't be added to the association's target. I don't want to suggest doing something that could change application behaviour.

I suppose this change could be a new framework default instead of a deprecation cycle, if you think giving users the ability to fully opt in (or not) is valuable.

kamipo replied:
I think this change is similar to https://github.com/ruby/ruby/pull/1506 (https://bugs.ruby-lang.org/issues/13083) in terms of just for consistency.

Personally I'd not prefer to make inconvenient changes (allow `relation.new`, but disallow `record = relation.new; ... do something ...; record.save`, but allow `relation.create`) just for consistency.

(We've asked matz to reconsider and then have reverted that change, by the way https://twitter.com/nalsh/status/1201787737566638080)

Can we find any value other than consistency by this deprecation?

Is it not enough to update the doc?

eugeneius replied:
Allowing `relation.create` was my mistake; it should eventually raise an error rather than returning a read-only record.

I think this change and https://github.com/ruby/ruby/pull/1506 are a bit different. In the Ruby case, the goal was language design consistency (only `=~` is tolerant of nils) that users probably don't care too much about; changing the behaviour only inconvenienced them. In this case, the user has marked the relation as read-only, and enforcing read-only behaviour consistently makes that feature more useful, even if some existing code relies on the inconsistent enforcement and has to be rewritten.

I see this change as closer to https://github.com/rails/rails/pull/39496: the user has configured some behaviour and we should apply it everywhere, not just in whichever APIs happened to consider it when they were implemented.

eugeneius replied:
I pushed an update that deprecates `create` and `create!` entirely, and adds a way to opt in to the new behaviour for `new` and `build`, silencing the deprecation.

kamipo replied:
> I see this change as closer to #39496: the user has configured some behaviour and we should apply it everywhere, not just in whichever APIs happened to consider it when they were implemented.

I don't know if you will agree, but let me explain, the objective of those (#39495, #39496, #39501, #39515, #39530, #39789, and later PRs) is not just for consistency.

We had been reported some issues (e.g. #30279, #37554) which are caused by the lack of alias resolution.

People (includes me) usually use `alias_attribute` to match the legacy names to the current system's (or Rails') convention. In its usage, we would expect aliased attributes to behave the same way as the original attributes.

The value of resolving the issue will probably be agreed.

On the other hand in this case, users who do `relation.new` are obviously aimed at saving, I couldn't think there is value in blocking the save of the users.

And also, if we deprecates `relation.create`, do you think `relation.insert` should also be deprecated?

eugeneius replied:
I definitely agree that resolving attribute aliases everywhere has value beyond consistency: it allows users to achieve something that they couldn't before. What I meant is that I think this change does that too, I'll try to explain below.

> users who do `relation.new` are obviously aimed at saving, I couldn't think there is value in blocking the save of the users.

`relation.first.save` is also obviously an attempt to save a record, but we prioritise the intent of the user who defined the relation over that of the user interacting with it. The value of marking a relation as `readonly` is that it restricts the operations that can be performed on it; users who don't want these restrictions can remove the readonly option from the relation itself, or override it as necessary with `readonly(false)`.

On a large project, declaring a relation or association as `readonly` communicates to other developers or teams that the data should not be modified. Obviously it's still possible, with `readonly(false)` or `Model.where(id: relation.ids)` or `connection.execute`, but the initial `ReadOnlyRecord` error should prompt the user to reconsider whether they're doing something that wasn't intended.

> do you think `relation.insert` should also be deprecated?

Yes, along with `update_all` and `delete_all` (`destroy_all` already raises). Conceptually, I think a readonly relation should emulate a database view, in that no writes should be possible. I'm happy to add these deprecations, but I don't intend to merge this PR unless I can convince you, so I won't write the code yet.

rafaelfranca replied:
I support this change and I think it is close to the strict loading feature. I'd prefer if we make this the new default and just use the warning to let users know we are changing to the new behavior. If they called `readonly` they should not be able to create records in that relation/association.

eugeneius replied:
I'll rebase this and make those changes. Do you want to get this into 6.1, or should I bump the deprecation horizon?

rafaelfranca replied:
I think it is too late for 6.1, let's change the horizon

eugeneius committed:
Deprecate building saveable records from read-only relations

Currently records loaded from read-only relations are read-only, but
records built from them are not. It shouldn't be possible to modify the
data underlying a read-only relation, including adding new records.
